### PR TITLE
Sharapov Georgiy. Lab 3. Option 1.

### PR DESCRIPTION
**Pull request by nol0n**
Added optimization pass that replaces mul and add operation with muladd instruction.

### CODE REVIEW

File: llvm/test/lab3/sharapov_georgiy/test.mir [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+#     __m128d res1 = _mm_setzero_pd();
+#     __m128d res2 = _mm_setzero_pd();
+#     __m128d res3 = _mm_setzero_pd();
+# 
+#     tmp1 = a2 * a3; 
+#     tmp2 = a1 * a2; 
+# 
+#     tmp3 = a3 * a1; 
+#     __m128d smth = tmp3 / a3; 
+#     res3 = tmp3 + a3;
```
Comment (m-ly4) [inline llvm/test/lab3/sharapov_georgiy/test.mir:85]: Could you please check the following modified case?
```suggestion
#     res3 = tmp3 + a3;
#     __m128d smth = tmp3 / a3; 
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/lab3/sharapov_georgiy/test.mir (hunk 1 above)
Comment (nol0n) [reply llvm/test/lab3/sharapov_georgiy/test.mir:85]: I added the suggested test and made changes in the pass for such cases.

**Review by m-ly4 — APPROVED** [review]

**Review by m-ly4 — CHANGES_REQUESTED** [review]
Could you please rename labs dir into lab3?

File: llvm/lib/Target/X86/CMakeLists.txt [hunk 2]
```diff
@@ -118,3 +118,6 @@ add_subdirectory(Disassembler)
 add_subdirectory(MCA)
 add_subdirectory(MCTargetDesc)
 add_subdirectory(TargetInfo)
+add_subdirectory(labs)
```
Comment (m-ly4) [inline llvm/lib/Target/X86/CMakeLists.txt]:
```suggestion
add_subdirectory(lab3)
```

File: llvm/lib/CMakeLists.txt [hunk 3]
```diff
@@ -66,3 +66,5 @@ endif()
 # Component post-processing
 LLVMBuildResolveComponentsLink()
 LLVMBuildGenerateCFragment(OUTPUT ${LLVMCONFIGLIBRARYDEPENDENCIESINC})
+
+set(LLVM_TEST_DEPENDS ${LLVM_TEST_DEPENDS} PARENT_SCOPE)
```
Comment (m-ly4) [inline llvm/lib/CMakeLists.txt:70]: Is this change really required?

File: llvm/lib/Target/CMakeLists.txt [hunk 4]
```diff
@@ -41,3 +41,5 @@ endforeach()
 # logic in llvm_map_components_to_libnames is order dependent on the target
 # libraries being created.
 set_property(GLOBAL PROPERTY LLVM_TARGETS_CONFIGURED On)
+
+set(LLVM_TEST_DEPENDS ${LLVM_TEST_DEPENDS} PARENT_SCOPE)
```
Comment (m-ly4) [inline llvm/lib/Target/CMakeLists.txt:45]: Is this change really required?

File: llvm/lib/Target/X86/CMakeLists.txt [hunk 5]
```diff
@@ -118,3 +118,6 @@ add_subdirectory(Disassembler)
 add_subdirectory(MCA)
 add_subdirectory(MCTargetDesc)
 add_subdirectory(TargetInfo)
+add_subdirectory(lab3)
+
+set(LLVM_TEST_DEPENDS ${LLVM_TEST_DEPENDS} PARENT_SCOPE)
```
Comment (m-ly4) [inline llvm/lib/Target/X86/CMakeLists.txt:123]: Is this change really required?

Comment (nol0n) [reply llvm/lib/Target/X86/CMakeLists.txt:123]: At least as far as I understand, yes, otherwise plugin won't compile, I tried. We done the same in all previous labs. I don't know how Artem was able to build it without this dependency and he doesn't know either.

**Review by m-ly4 — APPROVED** [review]
